### ISSUE TITLE
No definitions found for certain scenario

### ISSUE THREAD

**Issue by dmalyuta** [davidhalter/jedi#880]
Please see a detailed description and discussion in tkf/emacs-jedi#281, which turns out to be probably a Jedi issue. In brief:

1. Clone [this repository](https://github.com/ethz-asl/kalibr): `git clone https://github.com/ethz-asl/kalibr`
2. Go to [line 49](https://github.com/ethz-asl/kalibr/blob/master/aslam_offline_calibration/kalibr/python/kalibr_imu_camera_calibration/IccCalibrator.py#L49) and try to find definitions of `addDesignVariables`. No definitions are found, yet the `addDesignVariables` function is defined in [IccSensors.py](https://github.com/ethz-asl/kalibr/blob/master/aslam_offline_calibration/kalibr/python/kalibr_imu_camera_calibration/IccSensors.py#L316), which is a python file in the same directory.

Here is a Python console session showing the fault (courtesy [syohex](https://github.com/syohex) - which shows that I'm not the only one affected):
```
% ipython
Python 2.7.10 (default, Oct 23 2015, 19:19:21)
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import jedi

In [2]: source = open('IccCalibrator.py', 'r').read()

In [3]: script = jedi.Script(source, 49, len('        self.CameraChain.'), 'IccCalibrator.py')

In [4]: defs = script.goto_definitions()

In [5]: defs
Out[5]: []
```

I am running Ubuntu 14.04 and:
```
$ pip show jedi
---
Name: jedi
Version: 0.10.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: 
$ pip show python
---
Name: Python
Version: 2.7
Location: /usr/lib/python2.7/lib-dynload
Requires: 
```

Since [line 177](https://github.com/ethz-asl/kalibr/blob/master/aslam_offline_calibration/kalibr/python/kalibr_calibrate_imu_camera#L177) calls the [registerCamChain](https://github.com/ethz-asl/kalibr/blob/master/aslam_offline_calibration/kalibr/python/kalibr_imu_camera_calibration/IccCalibrator.py#L58) function with `camChain` (an `IccCameraChain` object) argument, Jedi should make the guess that `self.CameraChain` is an `IccCameraChain` object and thus should be able to find the definition of `addDesignVariables` on [line 49](https://github.com/ethz-asl/kalibr/blob/master/aslam_offline_calibration/kalibr/python/kalibr_imu_camera_calibration/IccCalibrator.py#L49), which is a method of the `IccCameraChain` class.

**Comment by davidhalter** [comment]
> No definitions are found, yet the addDesignVariables function is defined in IccSensors.py, which is a python file in the same directory.

The problem is not that it wouldn't find `addDesignVariables`. The problem is that it doesn't find the execution of `registerCamChain`.

**Comment by dmalyuta** [comment]
Why did you close this? In the last paragraph of my original post, I say: [line 177](https://github.com/ethz-asl/kalibr/blob/master/aslam_offline_calibration/kalibr/python/kalibr_calibrate_imu_camera#L177) calls the [registerCamChain](https://github.com/ethz-asl/kalibr/blob/master/aslam_offline_calibration/kalibr/python/kalibr_imu_camera_calibration/IccCalibrator.py#L58) function with `camChain` (an `IccCameraChain` object) argument. So it should be finding the execution of `registerCamChain`, right?

**Comment by davidhalter** [comment]
Hmm ok. I see. I still think that this is quite hard to understand. I'm reopening. However inheritance is hard to understand, if you want to do it right.

IMO this kind of code can be 

Sorry for closing prematurely and not reading your issue fully, because the description was actually good. I just tend to close aggressively to keep the tracker clean (and my limited time resources down).

**Comment by dmalyuta** [comment]
> IMO this kind of code can be

Could you please finish your sentence here?

Here is a simpler example of the same problem, where Jedi fails to find the definition based on context elsewhere in the project of where/with what type of arguments the function is called:
```
.
├── foo_module
│   ├── __init__.py
│   ├── MyCalibrator.py
│   ├── MySensor.py
│   └── MyUtil.py
└── main.py
```
\_\_init\_\_.py:
```
from MyCalibrator import *
import MyUtil as util
import MySensor as sens
```
MyCalibrator.py:
```
def Calibrate(sensor):
    sensor.calibrate()
```
MySensor.py:
```
import MyUtil as utils

class Sensor:
    def calibrate(self,arg):
        arg.Hello()
        print("Calibrating the sensor now!")
```
MyUtil.py:
```
def Hello():
    print("Hello World!")

class Bar:
    def Hello(self):
        print("Bar says hi!")
```
main.py:
```
from foo_module import *

util.Hello()

barObject=util.Bar()

sens.Sensor().calibrate(barObject)
```
Jedi fails to find definition of the `Hello` method in MySensor.py:
```
~/.../foo_module $ ipython
Python 2.7.6 (default, Oct 26 2016, 20:30:19) 
Type "copyright", "credits" or "license" for more information.

IPython 1.2.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import jedi

In [2]: source = open('MySensor.py','r').read()

In [3]: script = jedi.Script(source, 7, len('        arg.'), 'MySensor.py')

In [4]: defs = script.goto_definitions()

In [5]: defs
Out[5]: []

```

**Comment by davidhalter** [comment]
Sorry again.

IMO this kind of code can be understood, but I'm not sure it will be in the next one or two years. Understanding this form of inheritance is possible, but it's hard to implement and I have other plans currently. I'll leave it open however.

If anyone (you?) wants to look into this, the `jedi.evaluate.dynamic` module needs to be slightly changed to search for inherited class calls as well and not only calls to inherited classes.

~ Dave

**Comment by dmalyuta** [comment]
I see. Is this kind of functionality not really useful to professional Python developers (I'm new to the language), that it is something you don't see as being urgent to implement? What do you use when writing Python to quickly and efficiently call functions/methods which (like in this example) you can't autocomplete/immediately jump to definitions of?

**Comment by davidhalter** [comment]
Well, I guess for "professionals" there's other issues that are more important. Among the issues are lacking virtuelenv support, lacking Django support and so on. I think the case you mentioned is something that is not very typical for Python. I also think that it's not a good pattern to change the state in such a function (and therefore probably nobody ever had that issue before).

**Comment by dmalyuta** [comment]
> change the state in such a function 

What do you mean?

**Comment by davidhalter** [comment]
```
    def registerCamChain(self, sensor):
        self.CameraChain = sensor
```

This is usually a bad idea. It should be part of the instance creation call, IMO.

**Comment by davidhalter** [comment]
I realized that the only way we would ever fix that is by implementing #1059.
